### PR TITLE
fix issue in #reify_has_many_through

### PR DESCRIPTION
**Pull request by theRealNG**
Fix for https://github.com/airblade/paper_trail/issues/590

The assoc.foreign_key holds true only when the association is as follows

```ruby
Book
has_many :authorships
has_many :authors, :through => :authorships

Authorships
belongs_to :book
belongs_to :author

Author
has_many :authorships
has_many :books, :through => :authorships
```

In this case the assoc.foreign_key and assoc.association_foreign_key are the same

But in the case of the following associations

```ruby
Document
has_many :sections
has_many :paragraphs, :through => :sections

Section
belongs_to :document
has_many :paragraphs

Paragraph
belongs_to :section
```

the assoc.foreign_key and assoc.association_foreign_key are different and assoc.association_foreign_key gets the correct list of associated foreign_keys for the model.

For Ref: http://www.rubydoc.info/docs/rails/3.0.0/ActiveRecord%2FReflection%2FAssociationReflection%3Aassociation_foreign_key

EDIT: 
When a nested has_many relation is present, using the "reify_has_manys" method to reify the has_many through associations.

### CODE REVIEW

**Comment by jaredbeck** [comment]
Thanks for the PR!  Can you add a test that demonstrates the issue?

**Comment by jaredbeck** [comment]
> Can you add a test that demonstrates the issue?

Just to clarify, we know that #590 is a confirmed bug, but I'd still be nice to have a test.  Thanks.

**Comment by theRealNG** [comment]
Ya sure. Will work on it.

**Comment by jaredbeck** [comment]
You should be able to use the failing test I already wrote (https://github.com/jaredbeck/pt_issue_590/blob/de54892426966d146a13c8c2382735826aa76660/spec/model/document_spec.rb) as a starting point.

File: test/unit/model_test.rb [hunk 1]
```diff
@@ -1902,4 +1902,152 @@ class HasPaperTrailModelTransactionalTest < ActiveSupport::TestCase
...
+        @chapter.sections.create :name => "section 1"
+        Timecop.travel 1.second.since
+        @chapter.sections.first.update_attributes :name => "section 2"
+        Timecop.travel 1.second.since
+        @chapter.update_attributes :name => "ch_3"
+        Timecop.travel 1.second.since
+        @chapter.sections.first.update_attributes :name => "section 3"
+      end
+
+      context "when reified is called" do
```
Comment (jaredbeck) [inline test/unit/model_test.rb]: The method is called `reify`, not `reified`.  Please update the context string here.

File: test/unit/model_test.rb [hunk 2]
```diff
@@ -1902,4 +1902,152 @@ class HasPaperTrailModelTransactionalTest < ActiveSupport::TestCase
...
+        setup do
+          @chapter_1 = @chapter.versions.last.reify(:has_many => true)
+        end
+
+        should "show the value of the base record as it was before" do
+          assert_equal "ch_2", @chapter_1.name
```
Comment (jaredbeck) [inline test/unit/model_test.rb]: If the chapter name is "ch_2", then the instance variable here should be named `@chapter_2`, not `@chapter_1` as you have.

File: lib/paper_trail/version_concern.rb [hunk 3]
```diff
@@ -417,7 +417,14 @@ def reify_has_many_through(associations, model, options = {})
       associations.each do |assoc|
         next unless assoc.klass.paper_trail_enabled_for_model?
         through_collection = model.send(assoc.options[:through])
-        collection_keys = through_collection.map { |through_model| through_model.send(assoc.foreign_key) }
+
+        # if the association is a has_many association again, then call reify_has_manys for each through_collection
+        if !assoc.source_reflection.belongs_to? && through_collection.present?
+          through_collection.each { |through_model| reify_has_manys(through_model,options) }
+          return
```
Comment (jaredbeck) [inline lib/paper_trail/version_concern.rb]: Does this return `nil`?  That seems different from the existing return type, which is `associations`, right? We probably don't really care about the return value, so maybe we should just always return `nil`?

**Comment by jaredbeck** [comment]
@theRealNG this is looking good.  I appreciate how thorough you were with the tests.  My comments are mostly about style. I'm trying my best to understand the meat of the change, but I'd really like to get Ben's opinion as he is more familiar with the reify functions.

CC: @batter

File: test/unit/model_test.rb (hunk 2 above)
Comment (batter) [inline test/unit/model_test.rb]: Agreed, the way you are modifying the attributes here vs the variable names in usage is a bit confusing. There are other spots int he test suite where we have similar confusing attribute modification vs variable names, specifically within this file, but that doesn't mean we should continue to follow that pattern.

File: lib/paper_trail/version_concern.rb (hunk 3 above)
Comment (batter) [inline lib/paper_trail/version_concern.rb]: Should this really be returning at this point?

Comment (theRealNG) [inline lib/paper_trail/version_concern.rb]: Yes, So that the rest of the code is not executed. 
We will be getting the required records by calling "reify_has_manys".

**Comment by theRealNG** [comment]
Bug: 
We are not handling nested has_many associations.

Fix:
For a nested has_many associations we can call "has_manys" method for each of the nested associations so that we handle any levels of nesting. 

Explanation with example :- 
* A chapter has many sections and a section has many paragraphs.
* So when we call "reify_has_many_directly" on the chapter object, we get a list of sections that are associated with the chapter. 
* Since the association between section and paragraphs is also a has_many association, we can call reify_has_many_directly on each of the sections we got from the above result to get all the paragraphs associated to that section and in turn associated to the chapter. 

But i'm calling "reify_has_manys" on each section so that we can handle any levels of nesting.

File: lib/paper_trail/version_concern.rb (hunk 3 above)
Comment (batter) [inline lib/paper_trail/version_concern.rb]: I guess my question / thought is, shouldn't the method still perform the same behavior it does prior to the return in these cases? In the normal / existing method [flow we assign the new collection of updated issues to the association definition](https://github.com/theRealNG/paper_trail/blob/issue%23590/lib/paper_trail/version_concern.rb#L458) and return it in one call... In this case is it not necessary to perform that since the collection is already assigned to the association, or am I missing something here?

Comment (theRealNG) [inline lib/paper_trail/version_concern.rb]: The collection is assigned to the association in the "reify_has_many_directly" method also.

File: lib/paper_trail/version_concern.rb [hunk 4]
```diff
@@ -417,7 +417,14 @@ def reify_has_many_through(associations, model, options = {})
...
+        end
+
+        collection_keys = through_collection.map { |through_model| through_model.send(assoc.association_foreign_key)}
```
Comment (jaredbeck) [inline lib/paper_trail/version_concern.rb]: If I revert the change to this one line, by calling `assoc.foreign_key` instead of `assoc.association_foreign_key`, all the tests still pass.  So, is it necessary to change this line?

Comment (theRealNG) [inline lib/paper_trail/version_concern.rb]: In the example of Document, Sections and paragraphs:
The through_model is a object of Section.

What we are trying to do is get the list of all the paragraphs associated to the section by calling:
section.paragraph_ids

If we do Section.reflect_on_association(:paragraphs).foreign_key, it returns 'section_id' and if we do Section.reflect_on_association(:paragraphs).association_foreign_key, it returns 'paragraph_id'. So we should be using 'association_foreign_key' instead of 'foreign_key'.

This was not an issue so far because in a 'belongs_to' association assoc.association_foreign_key and assoc.foreign_key are the same.

Comment (jaredbeck) [inline lib/paper_trail/version_concern.rb]: That all makes sense, thanks.

File: lib/paper_trail/version_concern.rb (hunk 3 above)
Comment (batter) [inline lib/paper_trail/version_concern.rb]: I believe we want to use `next` instead of `return` here, because the code is executing through [a loop at this point and we don't want to prematurely return out of the method and break the loop](https://github.com/theRealNG/paper_trail/blob/issue%23590/lib/paper_trail/version_concern.rb#L417)

Comment (jaredbeck) [inline lib/paper_trail/version_concern.rb]: > I believe we want to use next instead of return here, because the code is executing through a loop at this point and we don't want to prematurely return out of the method and break the loop

I agree.  Here's a test that fails with `return`, passes with `next`: https://github.com/theRealNG/paper_trail/pull/1

Comment (theRealNG) [inline lib/paper_trail/version_concern.rb]: Yup @batter you are right. Thanks @jaredbeck

**Comment by jaredbeck** [comment]
Ben, I think this is ready to merge, now that it includes https://github.com/theRealNG/paper_trail/pull/1

What do you think?

**Comment by theRealNG** [comment]
Do you want me to squash the commits into a single commit?

**Comment by jaredbeck** [comment]
> Do you want me to squash the commits into a single commit?

Sure, thanks.

**Comment by batter** [comment]
:+1: Nice work, thanks to both of you.

My take is it's fine to leave it as a split commit as long as the merge has no fast forward but I don't care

**Comment by batter** [comment]
@jaredbeck Did the test we wrote ever end up working out of curiosity?

**Comment by jaredbeck** [comment]
> @jaredbeck Did the test we wrote ever end up working out of curiosity?

Yes, see https://github.com/theRealNG/paper_trail/pull/1

**Comment by batter** [comment]
So what is our merge process now for these types of things @jaredbeck?

Merge to `master`, then cherry-pick / backport to `4.1-stable`, `4.1-stable`?

**Comment by jaredbeck** [comment]
> So what is our merge process now for these types of things? 
> Merge to master, then cherry-pick / backport to [4.0-stable], 4.1-stable?

Yes, that's how I picture it working. I think it's a common strategy; it's what rails seems to do. One thing I'm not sure of is how to run CI for a backport. I guess we could do a PR for each backport..

**Comment by batter** [comment]
ok that sounds good, just wanted to be clear about how you were envisioning it
